### ISSUE TITLE
Sending file descriptors on unix sockets

### ISSUE THREAD

**Issue by webczat** [dotnet/runtime#932]
Hello,
I was wondering for some time: if unix sockets are supported directly by sockets api, how would I send unix file descriptors over them? It is a feature supported by every unix flavor, and pretty useful for example for implementing a dbus library, that comes to my mind first but it's not the only use case.
My question is, how would I do this with support for async/await? I am able to get a socket handle, so I could call sendmsg/recvmsg correctly, but I probably cannot attach to the io worker stuff in order to get non blocking notifications and implement SendFileDescriptorsAsync or something along the lines?
Also how would I go about implementing similar, more platform dependent features, like sending unix credentials or selinux context?

**Comment by wfurt** [comment]
You still should be able to use poll/select to know that you can read something. 
Question is what you do with descriptor passed to your process? 
We may be able to get https://github.com/dotnet/runtime/issues/862 but I don't know how would you for example use it with files.

**Comment by webczat** [comment]
generally I believe FileStream accepts handles, don't remember if SafeFileHandle or UIntPtr, but it probably does. in the first case not sure if SafeFileHandle accepted UIntPtr, would have to check, but I believe it was possible. Othervise not sure how all that systemd activation support works in kestrel, when it gets a raw fd from systemd? it is actually file descriptor passing I believe.
As for using poll/select, I could do that, but all the rest of socket methods that do async operations integrate with io workers that already do poll/select/epoll. it would be duplicating functionality wouldn't it?

**Comment by wfurt** [comment]
Yes, it seems like `SafeFileHandle` can be created from IntPtr. As far as the second part, are you talking about function on socket used for passing or about processing once new descriptor is received?

**Comment by webczat** [comment]
I mean the function on socket used for passing. I have to call sendmsg instead of send or recvmsg instead of recv depending on the context. in case of async send/receive it would do select/poll/epoll on linux wouldn't it? and call actual send/recv after notification. If I implement my own handling like ReceiveDescriptorsAsync/SendDescriptorsAsync (let's assume such naming) then suddenly I have to call different functions on notification, and i have no way of hooking up to the whole io worker thing.

**Comment by wfurt** [comment]
yes, that is correct. Since there is no way how to pass descriptor to/from PAL, you will have to p/invoke native functions directly.  I would probably use dedicated socket to pass auxiliary data and I would not mix it with any other use.

**Comment by webczat** [comment]
note that that way you directly force-restrict any use of such features for protocols you don't control. nothing known to me uses sockets in such a way that they exist only to pass aux data. especially when this data is not a fd, but for example something like process info/user credentials. These data are, as said, aux, so they are being send with something else, for example with a dbus message, or whatever. Another example would be libvirt, the linux virtualization api, passing file descriptors to an open display when you try to connect to your vm's console locally. I am aware of the fact that I have to use native functions directly, the only problem with me is that I cannot hook that logic up to the io worker stuff, so I need to actually reimplement unix sockets just for this feature, or there is any other way that I don't know... or, I just ignore async and do all synchronous at least for fd passing.

**Comment by wfurt** [comment]
yes, you need to implement IO functions as what you want is not part of official API contract. 
However you can still implement it in asynchronous way and get it scheduled on thread pool.
Once you know you can read/write data, there really is not that much difference between sync/async method (as even Async can finish synchronously)

**Comment by webczat** [comment]
won't it conflict with other async ops? also if I would like select/poll/epoll support I have problems as I have to reimplement that for fd passing...?
It would be useful to be able to attach to the io workers. It is already possible on windows because there is the mechanism of overlapped io and you have the ability to attach to this mechanism, but that is not possible here.

**Comment by karelz** [comment]
Triage: @antonfirsov please reconcile it with our DuplicateAndClose plans for Linux (related to #1760).

**Comment by webczat** [comment]
just wondering, if/until that is implemented, what is the expected course of action if I need this feature? I could invoke syscalls directly but async will never work. And there are potentially other similar things like sending credentials.

**Comment by wfurt** [comment]
triage: no resources in 8. We would need new API and we would need to check if there is any update on Windows side.
